node/pkg/watchers/ibc/watcher.go

```diff
@@ -0,0 +1,745 @@
+	})
+
+	// Start a routine for each chain to listen for observation requests.
+	for _, ce := range w.chainMap {
+		common.RunWithScissors(ctx, errC, "ibc_objs_req", func(ctx context.Context) error {
+			return w.handleObservationRequests(ctx, errC, ce)
+		})
+	}
+
+	// Signal to the supervisor that this runnable has finished initialization.
+	supervisor.Signal(ctx, supervisor.SignalHealthy)
+
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	case err := <-errC:
+		return err
+	}
+}
+
+// handleEvents reads messages from the IBC receiver contract and processes them.
+func (w *Watcher) handleEvents(ctx context.Context, c *websocket.Conn, errC chan error) error {
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case err := <-errC:
+			return fmt.Errorf("handleEvents died: %w", err)
+		default:
+			_, message, err := c.Read(ctx)
+			if err != nil {
+				w.logger.Error("failed to read socket", zap.Error(err))
+				connectionErrors.WithLabelValues("channel_read_error").Inc()
+				return fmt.Errorf("failed to read socket: %w", err)
+			}
+
+			// Received a message from the blockchain.
+			json := string(message)
+
+			txHashRaw := gjson.Get(json, "result.events.tx\\.hash.0")
+			if !txHashRaw.Exists() {
+				w.logger.Warn("message does not have tx hash", zap.String("payload", json))
+				continue
+			}
+			txHash, err := vaa.StringToHash(txHashRaw.String())
+			if err != nil {
+				w.logger.Warn("failed to parse txHash", zap.String("txHash", txHashRaw.String()), zap.Error(err))
+				continue
+			}
+
+			events := gjson.Get(json, "result.data.value.TxResult.result.events")
+			if !events.Exists() {
+				w.logger.Warn("message has no events", zap.String("payload", json))
+				continue
+			}
+
+			for _, event := range events.Array() {
+				if !event.IsObject() {
+					w.logger.Warn("event is invalid", zap.Stringer("tx_hash", txHash), zap.String("event", event.String()))
+					continue
+				}
+				eventType := gjson.Get(event.String(), "type").String()
+				if eventType == "wasm" {
+					evt, err := parseIbcReceivePublishEvent(w.logger, w.contractAddress, event, txHash)
+					if err != nil {
+						w.logger.Error("failed to parse wasm event", zap.Error(err), zap.String("event", event.String()))
+						continue
+					}
+
+					if evt != nil {
+						w.processIbcReceivePublishEvent(txHash, evt, "new")
+					}
+				} else {
+					w.logger.Debug("ignoring uninteresting event", zap.String("eventType", eventType))
+				}
+			}
+		}
+	}
+}
+
+// handleQueryBlockHeight gets the latest block height from wormchain each interval and updates the status on all the connected chains.
+func (w *Watcher) handleQueryBlockHeight(ctx context.Context, c *websocket.Conn, errC chan error) error {
+	const latestBlockURL = "blocks/latest"
+
+	t := time.NewTicker(5 * time.Second)
+	client := &http.Client{
+		Timeout: time.Second * 5,
+	}
+
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case err := <-errC:
```

Comment:
This does not look like the correct usage. The error channel is shared so only one routine will receive the error. Best to catch the error on the top routine and cancel the context.